rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.allocation_explain/10_basic.yml

```diff
@@ -92,3 +92,21 @@
   - is_true: can_rebalance_cluster
   - is_true: can_rebalance_to_other_node
   - is_true: rebalance_explanation
+
+---
+"Cluster allocation explanation response includes node's roles":
+  - skip:
+      version: " - 8.9.99"
+      reason: The roles field was introduced in 8.10.0
+
+  - do:
+      indices.create:
+        index: test
+
+  - match: { acknowledged: true }
+
+  - do:
+      cluster.allocation_explain:
+        body: { "index": "test", "shard": 0, "primary": true }
+
+  - is_true: current_node.roles
```

Comment:
Roles should be added to other places as well:
* allocate unassigned section (if present)
* move shard section (if present)
* each node of the per node explanation

I believe the first 2 are not present since this explain is executed against allocated shard that could stay on the current node. I do not think they require additional test case and could be covered by serialization test, see https://github.com/elastic/elasticsearch/pull/98550/files#r1299861004

I wonder about the last one. I think it could be covered with:
`- is_true: node_allocation_decisions.0.roles`
unless I miss something.

Reply:
The YAML tests (sometimes) run in a one-node cluster so we can't assert anything about the other changed locations here.

I think we could check the `node_allocation_decisions` bit in `ClusterAllocationExplainIT#testExplainRolesOutput` if you'd like, by starting a second node with different roles (after creating the index, so we know which node has the shard). I didn't see an easy way to get nodes in the `AllocateUnassignedDecision` or `MoveDecision` here, it might be easier to do that in a unit test, but tbh it's the `node_allocation_decision` that really matters.

Reply:
Extended `ClusterAllocationExplainIT#testExplainRolesOutput` to test `node_allocation_decisions` 👍